pkg/operator/sync_kubecontrollermanager_v311_00.go

```diff
@@ -28,66 +27,60 @@ func syncKubeControllerManager_v311_00_to_latest(c KubeControllerManagerOperator
-	_, _, err = resourceapply.ApplyService(c.corev1Client, requiredService)
-	if err != nil {
-		errors = append(errors, fmt.Errorf("%q: %v", "svc", err))
-	}
-
-	requiredSA := resourceread.ReadServiceAccountV1OrDie(v311_00_assets.MustAsset("v3.11.0/kube-controller-manager/sa.yaml"))
-	_, saModified, err := resourceapply.ApplyServiceAccount(c.corev1Client, requiredSA)
-	if err != nil {
-		errors = append(errors, fmt.Errorf("%q: %v", "sa", err))
-	}
-
-	controllerManagerConfig, configMapModified, err := manageKubeControllerManagerConfigMap_v311_00_to_latest(c.corev1Client, operatorConfig)
+	directResourceResults := resourceapply.ApplyDirectly(c.kubeClient, v311_00_assets.Asset,
+		"v3.11.0/kube-controller-manager/clusterrolebinding.yaml",
+		"v3.11.0/kube-controller-manager/ns.yaml",
+		"v3.11.0/kube-controller-manager/public-info-role.yaml",
+		"v3.11.0/kube-controller-manager/public-info-rolebinding.yaml",
+		"v3.11.0/kube-controller-manager/svc.yaml",
+		"v3.11.0/kube-controller-manager/sa.yaml",
+		"v3.11.0/openshift-apiserver/sa.yaml",
+	)
+	resourcesThatForceRedeployment := sets.NewString("v3.11.0/kube-controller-manager/sa.yaml")
+	forceRollingUpdate := false
+
+	for _, currResult := range directResourceResults {
+		if currResult.Error != nil {
+			errors = append(errors, fmt.Errorf("%q (%T): %v", currResult.File, currResult.Type, currResult.Error))
+			continue
+		}
+
+		if currResult.Changed && resourcesThatForceRedeployment.Has(currResult.File) {
+			forceRollingUpdate = true
+		}
+	}
+
+	controllerManagerConfig, configMapModified, err := manageKubeControllerManagerConfigMap_v311_00_to_latest(c.kubeClient.CoreV1(), operatorConfig)
 	if err != nil {
 		errors = append(errors, fmt.Errorf("%q: %v", "configmap", err))
 	}
 
 	forceRollout := operatorConfig.ObjectMeta.Generation != operatorConfig.Status.ObservedGeneration
-	if saModified { // SA modification can cause new tokens
-		forceRollout = true
-	}
-	if configMapModified {
-		forceRollout = true
-	}
+	forceRollout = forceRollout || forceRollingUpdate || configMapModified
 
 	// our configmaps and secrets are in order, now it is time to create the DS
 	// TODO check basic preconditions here
-	actualDeployment, _, err := manageKubeControllerManagerDeployment_v311_00_to_latest(c.appsv1Client, operatorConfig, previousAvailability, forceRollout)
+	actualDaemonSet, _, err := manageKubeControllerManagerDaemonSet_v311_00_to_latest(c.kubeClient.AppsV1(), operatorConfig, previousAvailability, forceRollout)
 	if err != nil {
-		errors = append(errors, fmt.Errorf("%q: %v", "deployment", err))
+		errors = append(errors, fmt.Errorf("%q: %v", "daemonset", err))
+	}
+
+	// if our actual daemonset has at least one pod available, then we should delete the kube-system daemonset to make sure we take over for it.
+	if actualDaemonSet.Status.NumberAvailable > 0 {
+		// we don't care about the return value here.
+		c.kubeClient.AppsV1().DaemonSets("kube-system").Delete("kube-controller-manager", nil)
```

Comment:
@sttts I really want this delete in order to be sure that our controller manager actually works (it didn't before).  Is it going to break bootkube?

Reply:
We move the phase 2 daemonset into the operator namespace. Then we can render exactly the manifest you would create in phase 3. So you can easily adopt it.